### PR TITLE
Openstack identity v3 builder using the new gophercloud API

### LINKED ISSUE #1626: Upgrade to latest GopherCloud API
This will likely fix a lot of bugs since we're pinned at a super old (July 2014) version right now. But we're going to remain pinned until we can figure out the upgrade process.

### PR DESCRIPTION
**Pull request by einarf**
Fix #1626 #480

This is an attempt to port the openstack builder to use the new gophercloud API so Packer can eventually work well for different kinds of openstack setups. It does exactly what the old builder is doing step by step. 

Throwing this in as a new builder for now, but I guess we have to figure out what to do with this thing. The intention of this builder was initially to get packer to work with Zetta.IO's openstack using identity v3, but it should in theory work just like the old builder. Needs testing.

It took some effort as the gophercloud API was missing some functionality (@ggiamarchi fixed missing part of id3 support and @einarf added image creation). The final PR just got merged into gophercloud.. so here we are..

The builder definitely runs and works.

### CODE REVIEW

**Comment by einarf** [comment]
Think this is pretty much done one. @mitchellh or anyone around?

File: builder/openstack_id3/access_config.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	c.Username = common.ChooseString(c.Username, os.Getenv("SDK_USERNAME"), os.Getenv("OS_USERNAME"))
+	c.UserId = common.ChooseString(c.UserId, os.Getenv("OS_USER_ID"))
+	c.Password = common.ChooseString(c.Password, os.Getenv("SDK_PASSWORD"), os.Getenv("OS_PASSWORD"))
+	c.ApiKey = common.ChooseString(c.ApiKey, os.Getenv("SDK_API_KEY"))
+	c.Project = common.ChooseString(c.Project, os.Getenv("SDK_PROJECT"), os.Getenv("OS_TENANT_NAME"), os.Getenv("OS_PROJECT_NAME"))
+	c.Provider = common.ChooseString(c.Provider, os.Getenv("SDK_PROVIDER"), os.Getenv("OS_AUTH_URL"))
+	c.RawRegion = common.ChooseString(c.RawRegion, os.Getenv("SDK_REGION"), os.Getenv("OS_REGION_NAME"))
+	c.TenantId = common.ChooseString(c.TenantId, c.ProjectId, os.Getenv("OS_TENANT_ID"), os.Getenv("OS_PROJECT_ID"))
+	c.Domain = common.ChooseString(c.Domain, os.Getenv("OS_DOMAIN_NAME"))
+	c.DomainId = common.ChooseString(c.DomainId, os.Getenv("OS_PROJECT_DOMAIN_ID"), os.Getenv("OS_USER_DOMAIN_ID"))
```
Comment (sethvargo) [inline builder/openstack_id3/access_config.go]: Should we validate the existence of some of these options (i.e. the required ones) or is that handled by gophercloud?

File: builder/openstack_id3/access_config.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		return nil, err
+	}
+	return provider, nil
+}
+
+func (c *AccessConfig) Region() string {
+	return common.ChooseString(c.RawRegion, os.Getenv("SDK_REGION"), os.Getenv("OS_REGION_NAME"))
+}
+
+func (c *AccessConfig) Prepare(t *packer.ConfigTemplate) []error {
```
Comment (sethvargo) [inline builder/openstack_id3/access_config.go]: I think this should probably be a [multierror](https://github.com/hashicorp/go-multierror). What do you think?

File: builder/openstack_id3/access_config.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	// Attempts to autheticate and update provider with token and catalog
+	err = openstack.Authenticate(provider, authoptions)
+	if err != nil {
+		return nil, err
+	}
+	return provider, nil
+}
+
+func (c *AccessConfig) Region() string {
+	return common.ChooseString(c.RawRegion, os.Getenv("SDK_REGION"), os.Getenv("OS_REGION_NAME"))
```
Comment (sethvargo) [inline builder/openstack_id3/access_config.go]: What happens if an envvar is changed between calls?

File: builder/openstack_id3/run_config.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		c.SSHPort = 22
+	}
+
+	if c.RawSSHTimeout == "" {
+		c.RawSSHTimeout = "5m"
+	}
+
+	if c.UseFloatingIp && c.FloatingIpPool == "" {
+		c.FloatingIpPool = "public"
+	}
```
Comment (sethvargo) [inline builder/openstack_id3/run_config.go]: Do you think it makes sense to move these into a `DefaultConfig()` function and then merge?

File: builder/openstack_id3/server.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+// an openstack server.
+func ServerStateRefreshFunc(sc *gophercloud.ServiceClient, s *servers.Server) StateRefreshFunc {
+	return func() (interface{}, string, int, error) {
+		server, err := servers.Get(sc, s.ID).Extract()
+		if err != nil {
+			urce, ok := err.(*perigee.UnexpectedResponseCodeError)
+			if ok && (urce.Actual == 404) {
+				log.Printf("404 on ServerStateRefresh, returning DELETED")
+
+				return nil, "DELETED", 0, nil
```
Comment (sethvargo) [inline builder/openstack_id3/server.go]: Should this be a constant?

File: builder/openstack_id3/access_config.go (hunk 1 above)
Comment (einarf) [inline builder/openstack_id3/access_config.go]: Gophercloud will validate them based on the identity version found in the auth endpoint.

**Comment by einarf** [comment]
The other things you pointed out are 100% identical to the current openstack builder. I can take a look, but my time is fairly limited.

File: builder/openstack_id3/access_config.go (hunk 3 above)
Comment (einarf) [inline builder/openstack_id3/access_config.go]: Looks like this one was made for the rackspace specific region check in `Prepare`, so we should probably not keep using it in the builder itself. Cleaning that up.

**Comment by sethvargo** [comment]
@einarf okay that's fine. I was doing a fresh review, but we can clean some of those things up later.

**Comment by einarf** [comment]
@sethvargo Great. Probably easier that way. Also not entirely sure if this should even be a separate builder as it should work just like the old openstack builder, just using new gophercloud and support for identity v2 and v3. I have not personally tested it for v2 though :)

The builder name "openstack-id3" can be misleading...

**Comment by einarf** [comment]
I think this one is done for now. What might remain is what this builder should be called.

**Comment by einarf** [comment]
@sethvargo anything I can do to get any progress on this PR?

File: builder/openstack_id3/ssh.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	return func(state multistep.StateBag) (string, error) {
+
+		s := state.Get("server").(*servers.Server)
+
+		if ip := state.Get("access_ip").(*floatingip.FloatingIP); ip.IP != "" {
+			return fmt.Sprintf("%s:%d", ip.IP, port), nil
+		}
+
+		// Does the pool actually exist?
+		if pool, ok := s.Addresses[sshinterface]; ok {
```
Comment (droyo) [inline builder/openstack_id3/ssh.go]: This does not replicate the behavior of the old 'openstack' builder when 'ssh_interface' is unset (""). The old builder would loop through all addresses and choose the first available one.

File: builder/openstack_id3/step_allocate_ip.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		if err != nil {
+			err := fmt.Errorf("Error creating floating ip from pool '%s'", s.FloatingIpPool)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+		instanceIp = newIp
+		ui.Say(fmt.Sprintf("Created temporary floating IP %s...", instanceIp.IP))
+	}
+	if instanceIp.IP != "" {
```
Comment (droyo) [inline builder/openstack_id3/step_allocate_ip.go]: This fails with a nil pointer dereference when not using floating IPs. I was able to fix this by making instanceIP a floatingip.FloatingIP (as opposed to a pointer), like the old builder did.

**Comment by droyo** [comment]
@einarf After patching the issues I commented on above, I was able to build an image against the v2 API just fine. I've opened https://github.com/ZettaIO/packer/pull/1 with my changes. Let me know if there is a better way to get these changes to you.

**Comment by einarf** [comment]
@droyo Thanks. I'll get it in here asap. :)

**Comment by mitchellh** [comment]
Supersceded by #2218. Sorry to jump in here, we felt it woudl be easier to just greenfield this with the GopherCloud API. We use your PR as a base for a lot of it. Thanks!

**Comment by einarf** [comment]
All good! As long as it was useful in some way to get things done I am more than happy. Half of the work was in gophercloud itself.

**Pull request closed without merging**